File: examples/05-production/go/get_version_patch.go
package production

import (
	"time"

	"go.temporal.io/sdk/workflow"
)

// reserve and charge are Activities registered elsewhere.

// VersionedWorkflow shows the Go equivalent of Java's Workflow.getVersion:
// workflow.GetVersion returns DefaultVersion when replaying histories recorded
// before the change, and the new version for fresh executions, so old runs keep
// their original command order.
func VersionedWorkflow(ctx workflow.Context, orderID string) error {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
	})

	version := workflow.GetVersion(ctx, "charge-before-reserve", workflow.DefaultVersion, 1)
	if version == workflow.DefaultVersion {
		if err := workflow.ExecuteActivity(ctx, "Reserve", orderID).Get(ctx, nil); err != nil {
			return err
		}
		return workflow.ExecuteActivity(ctx, "Charge", orderID).Get(ctx, nil)
	}

	if err := workflow.ExecuteActivity(ctx, "Charge", orderID).Get(ctx, nil); err != nil {
		return err
	}
	return workflow.ExecuteActivity(ctx, "Reserve", orderID).Get(ctx, nil)
}

File: examples/05-production/python/get_version_patch.py
from datetime import timedelta

from temporalio import workflow

# `reserve` and `charge` are module-level @activity.defn functions.


@workflow.defn
class VersionedWorkflow:
    # workflow.patched / workflow.deprecate_patch are the Python equivalent of
    # Java's Workflow.getVersion. patched(id) returns True for new executions and
    # False when replaying histories recorded before the patch existed.
    @workflow.run
    async def run(self, order_id: str) -> None:
        if workflow.patched("charge-before-reserve"):
            await workflow.execute_activity(
                charge, order_id, start_to_close_timeout=timedelta(seconds=30)
            )
            await workflow.execute_activity(
                reserve, order_id, start_to_close_timeout=timedelta(seconds=30)
            )
        else:
            await workflow.execute_activity(
                reserve, order_id, start_to_close_timeout=timedelta(seconds=30)
            )
            await workflow.execute_activity(
                charge, order_id, start_to_close_timeout=timedelta(seconds=30)
            )
